pkg/core/native/native_gas.go

```diff
@@ -54,13 +54,20 @@ func newGAS(init int64) *GAS {
 	return g
 }
 
-func (g *GAS) increaseBalance(_ *interop.Context, _ util.Uint160, si *state.StorageItem, amount *big.Int) error {
+func (g *GAS) increaseBalance(_ *interop.Context, _ util.Uint160, si *state.StorageItem, amount *big.Int, checkBal *big.Int) error {
 	acc, err := state.NEP17BalanceFromBytes(*si)
 	if err != nil {
 		return err
 	}
 	if sign := amount.Sign(); sign == 0 {
-		return nil
+		// Requested self-transfer amount can be higher than actual balance.
+		if checkBal != nil {
+			if acc.Balance.Cmp(checkBal) < 0 {
+				err = errors.New("insufficient funds")
+			}
+		}
+		*si = acc.Bytes()
```

Comment:
Why do we need this line?
```
*si = acc.Bytes()
```

Reply:
Yeah, I've noticed that NEO does it in `increaseBalance`, but I've also questioned myself on why it does it. Zero balance? But it shouldn't be stored anyway. It needs to be checked, maybe it's not needed.

Reply:
It was added with 7987cdadc, so it's important for NEO, but it actually can be omitted for GAS (as it only stores balance).

Reply:
Fixed.